WIP: Add Doozer verb to build images with Konflux

vfreex opened the pull request.

ashwindasr commented on doozer/doozerlib/backend/konflux_image_builder.py:
```diff
@@ -0,0 +1,365 @@
+                "name": name,
+                "annotations": {
+                    "build.appstudio.openshift.io/pipeline": '{"name":"docker-build","bundle":"latest"}',
+                }
```
```
mintmaker.appstudio.redhat.com/disabled=true
```
Need to add this annotation in, just to be safe. ref [thread](https://redhat-internal.slack.com/archives/GDBRP5YJH/p1724260588819889?thread_ts=1723123622.362349&cid=GDBRP5YJH)

ashwindasr commented on doozer/doozerlib/backend/konflux_image_builder.py:
```diff
@@ -0,0 +1,365 @@
+        self.pipelinerun = pipelinerun
+
+
+KONFLUX_BUNDLE_PULLSPEC = "quay.io/konflux-ci/tekton-catalog/pipeline-docker-build:f3ac40bbc0230eccb8d98a4d54dabd55a4943c5d"
```
Looks like that SHA is pretty outdated. We probably need to figure out a process, but for now we can probably use the latest build always. But I see that there is `devel` so can we try something like

`quay.io/konflux-ci/tekton-catalog/pipeline-docker-build:devel`

thegreyd commented on doozer/doozerlib/backend/konflux_image_builder.py:
```diff
@@ -0,0 +1,319 @@
+        component_name = f"{app_name}-{metadata.distgit_key}"
+        dest_image_repo = self._config.output_repo
+        # dest_image_tag = f"{metadata.distgit_key}_{self._uuid_tag}"
+        dest_image_tag = f"{version}-{release}"
```
why not name-version-release?

vfreex replied:
You are right. Initially by default config each image gets an separate repo. Since we are going to use a mono repo for all images, we need to include the image name.

thegreyd commented on doozer/doozerlib/backend/konflux_image_builder.py:
```diff
@@ -0,0 +1,319 @@
+
+        })
+        obj = yaml.load(rendered)
+        del obj["metadata"]["name"]
+        del obj["metadata"]["namespace"]
+        del obj["metadata"]["annotations"]["pipelinesascode.tekton.dev/on-cel-expression"]
```
curious what's the reason behind this / or why do we have this in yaml in the first place if we don't want them?

vfreex replied:
In the future the PipelineRun template file (https://github.com/openshift-eng/art-tools/pull/859/files#diff-f1894aed6cfd6dbe45cd9701265dbf035239886d026b4dca0c40865f5e8511fe) will be loaded from an external repo, which includes some fields that are specific to a single image. We'd better to remove them.

vfreex replied:
I'll add a comment here

thegreyd commented on doozer/doozerlib/backend/konflux_image_builder.py:
```diff
@@ -0,0 +1,319 @@
+            else:
+                self._logger.warning(f"[DRY RUN] Would have created {api_version}/{kind} {namespace}/{name}")
+            return resource.ResourceInstance(dyn_client, manifest)
+        if found:
```
```suggestion
        if found:
            self._logger.info(f"Patching {api_version}/{kind} {namespace}/{name}")
```

thegreyd commented on doozer/doozerlib/backend/konflux_image_builder.py:
```diff
@@ -0,0 +1,319 @@
+                namespace=namespace,
+                content_type="application/merge-patch+json",
+            )
+        else:
```
```suggestion
        else:
            self._logger.info(f"Creating {api_version}/{kind} {namespace}/{name}")
```

thegreyd commented on doozer/doozerlib/backend/rebaser.py:
```diff
@@ -0,0 +1,1549 @@
+            sb.write(self._runtime.assembly)
+
+        el_ver = 0
+        try:
+            el_ver = metadata.branch_el_target()
+        except ValueError:
+            pass
+        if el_ver:
```
We've grown to depend on `.elX` suffix in our nvr names, so i think we should error out if it is not found instead of passing, so we can investigate

thegreyd replied:
but can also do in a separate PR, so upto you if you wanna make this change here

vfreex replied:
Right. This part was taken from existing code and I didn't change it. Let's make a change in future PRs.

thegreyd commented on doozer/doozerlib/backend/rebaser.py:
```diff
@@ -0,0 +1,1549 @@
+        dfp.add_lines(
+            "\n# Start Konflux-specific steps",
+            "RUN mkdir -p /tmp/yum_temp; mv /etc/yum.repos.d/*.repo /tmp/yum_temp/ || true",
+            "COPY .oit/signed.repo /etc/yum.repos.d/",
```
I think we want to change this to be signed/unsigned depending on the version 

https://github.com/openshift-eng/art-tools/blob/24fd4652d8ef4b803b28c21c3d0d1f9aa6d21340/doozer/doozerlib/osbs2_builder.py#L161

vfreex replied:
This part was taken from original Ashwin's PR. Let's address this in a separate PR.

ashwindasr commented on doozer/doozerlib/constants.py:
```diff
@@ -29,6 +29,12 @@
 KONFLUX_REPO_CA_BUNDLE_TMP_PATH = "/tmp"
 KONFLUX_REPO_CA_BUNDLE_FILENAME = "Current-IT-Root-CAs.pem"
 KONFLUX_REPO_CA_BUNDLE_HOST = "https://certs.corp.redhat.com/certs"
+WORKING_SUBDIR_KONFLUX_BUILD_SOURCES = "konflux_build_sources"
```
Noticed that these constants aren't being used? Or not sure if I missed something or it'll be in an another PR.

vfreex replied:
right, it is used in https://github.com/openshift-eng/art-tools/pull/820

vfreex commented:
@thegreyd @ashwindasr Thanks for the review. I've addressed your review suggestions. Please have another look.
Note previously this PR includes some commits from the rebase PR https://github.com/openshift-eng/art-tools/pull/820, so it was a bit confusing. For those review suggestions, I will open another PR to address them.

thegreyd commented:
/lgtm
